### PR TITLE
usb: device_next: changes around descriptor handling, preparation for the USB device class skeleton

### LINKED ISSUE #58168: usbd_shell: Conflict USB devices with user-defined USB device
**Describe the bug**
The USBD_DEVICE_DEFINE defined in zephyr/subsys/usb/device_next/usbd_shell.c will conflict the user-defined USB device
Thereby affecting the enumeration of user-defined USB devices

https://github.com/zephyrproject-rtos/zephyr/blob/91ffb72c3b07338f5f001d98b51417fa59d53a62/subsys/usb/device_next/usbd_shell.c#LL245C1-L245C1

the usbd_event_handler will be trigger twice

```
static void usbd_thread(void)
{
	struct udc_event event;

	while (true) {
		k_msgq_get(&usbd_msgq, &event, K_FOREVER);

		STRUCT_SECTION_FOREACH(usbd_contex, uds_ctx) {
			if (uds_ctx->dev == event.dev) {
				usbd_event_handler(uds_ctx, &event);
			}
		}
	}
}
```

### PR DESCRIPTION
**Pull request by jfischer-no**
Changes around descriptor handling, preparation for the USB device class skeleton

Commit 15b8df1edebf07020ee2e2b2c672267c10e7daca fixes: #58168

### CODE REVIEW

File: include/zephyr/usb/usbd.h [hunk 1]
```diff
@@ -62,7 +62,7 @@ extern "C" {
  */
 struct usbd_desc_node {
 	/** slist node struct */
-	sys_snode_t node;
+	sys_dnode_t node;
```
Comment (tmon-nordic) [inline include/zephyr/usb/usbd.h:70]: Where are you using the features that dlist has that are not found in slist? The commit message mentions "next commit" but it is not in next commit and even in the last commit I didn't find (maybe I just missed) where the dlist is really needed over slist.

File: subsys/usb/device_next/usbd_core.c [hunk 2]
```diff
@@ -199,6 +201,23 @@ int usbd_device_init_core(struct usbd_contex *const uds_ctx)
 
 int usbd_device_shutdown_core(struct usbd_contex *const uds_ctx)
 {
+	struct usbd_config_node *cfg_nd;
+	int ret;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&uds_ctx->configs, cfg_nd, node) {
+		uint8_t cfg_value = usbd_config_get_value(cfg_nd);
+
+		ret = usbd_class_remove_all(uds_ctx, cfg_value);
```
Comment (tmon-nordic) [inline subsys/usb/device_next/usbd_core.c:210]: I think the commits are overly separated. I would prefer if the helpers were added and used in the same commit.

Comment (jfischer-no) [reply subsys/usb/device_next/usbd_core.c:210]: They actually are, https://github.com/zephyrproject-rtos/zephyr/pull/58187/commits/6e1fdff22981fdcbcdec6fe91d57df7e1b14f59d. Do you think I should squash it with the last commit?

Comment (tmon-nordic) [reply subsys/usb/device_next/usbd_core.c:210]: I think one commit for the helpers + use helpers, keep the last commit separate.

**Review by tmon-nordic — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**